dive_core/stl_replacement.hpp

```diff
@@ -0,0 +1,267 @@
+    a.m_reserved = 0;
+    a.m_size = 0;
+    return *this;
+}
+
+//--------------------------------------------------------------------------------------------------
+template<class Type> Type *Vector<Type>::data() const
+{
+    return m_buffer;
+}
+
+//--------------------------------------------------------------------------------------------------
+template<class Type> Type &Vector<Type>::front() const
+{
+    return *m_buffer;
+}
+
+//--------------------------------------------------------------------------------------------------
+template<class Type> Type &Vector<Type>::back() const
+{
+    return m_buffer[m_size - 1];
+}
+
+//--------------------------------------------------------------------------------------------------
+template<class Type> uint64_t Vector<Type>::size() const
+{
+    return m_size;
+}
+
+//--------------------------------------------------------------------------------------------------
+template<class Type> uint64_t Vector<Type>::capacity() const
+{
+    return m_reserved;
+}
+
+//--------------------------------------------------------------------------------------------------
+template<class Type> bool Vector<Type>::empty() const
+{
+    return m_size == 0;
+}
+
+//--------------------------------------------------------------------------------------------------
+template<class Type> void Vector<Type>::push_back(const Type &a)
+{
+    reserve(m_size + 1);
+    new (&m_buffer[m_size]) Type(a);
+    ++m_size;
+}
+
+//--------------------------------------------------------------------------------------------------
+template<class Type> void Vector<Type>::push_back(Type &&a)
+{
+    emplace_back(std::move(a));
+}
+
+//--------------------------------------------------------------------------------------------------
+template<class Type> void Vector<Type>::pop_back()
+{
+    if (m_size > 0)
+    {
+        m_buffer[m_size - 1].~Type();
+        --m_size;
+    }
+}
+
+//--------------------------------------------------------------------------------------------------
+template<class Type> template<typename... Args> void Vector<Type>::emplace_back(Args &&...args)
+{
+    reserve(m_size + 1);
+
+    // Construct the element in place
+    new (m_buffer + m_size) Type(std::forward<Args>(args)...);
+    ++m_size;
+}
+
+//--------------------------------------------------------------------------------------------------
+template<class Type> void Vector<Type>::resize(uint64_t size)
+{
+    reserve(size);
+
+    // Call the constructor for all the new elements
+    for (uint64_t i = m_size; i < size; ++i)
+        new (&m_buffer[i]) Type();
+    m_size = size;
+}
+
+//--------------------------------------------------------------------------------------------------
+template<class Type> void Vector<Type>::resize(uint64_t size, const Type &a)
+{
+    reserve(size);
+
+    // Call the copy constructor for all the new elements
+    for (uint64_t i = m_size; i < size; ++i)
+        new (&m_buffer[i]) Type(a);
```

Comment:
not sure how much better it is, but you may skip calling the constructor if the type is a simple type by checking `std::is_trivially_constructible`